cms/djangoapps/contentstore/asset_storage_handlers.py

```diff
@@ -88,6 +88,52 @@ def handle_assets(request, course_key_string=None, asset_key_string=None):
     return HttpResponseNotFound()
 
 
+def check_asset_pre_upload(request, course_key_string):
+    """
+    Check if the files have any filename conflicts with existing assets.
+
+    Returns a JSON response with the following fields:
+    - allowed: boolean, whether the user is allowed to upload an asset
+    - reason: string, the reason why the user is not allowed to upload an asset
+    - files: list of strings, the filenames that already exist
+    """
```

Comment:
It looks like this endpoint is designed to become a place where we can add additional checks other than just name conflict checking. It also looks like it's making decisions about what to allow and what not to allow (with the 409 error). But this endpoint doesn't really make that decision–the person uploading will be allowed to overwrite if they want to. This endpoint exists to better inform their decision to make sure they're overwriting intentionally.

Given that, maybe it would be better to model this as an endpoint that just returns information about a set of filenames? So say we still POST up a bunch of filenames, and it gives us a few bits of basic information based on the metadata we get back in the contentstore `find` call? Something like:

```javascript
{
  "files": {
    "handout.pdf": {
      "size": 20930,
      "hash": "938c2cc0dcc05f2b68c4287040cfcf71"
    },
    "non-existant-handout.pdf": null
  }
}
```

It could then be a more generically named view that could possibly be used in circumstances other just uploads.

Non-existent files return with `null`, and the frontend code can decide what to do with that information–abort, upload it anyway, etc. If we want to add more metadata later to do more clever things, we can–e.g. check the hash to make sure we're not overwriting with the same file (out of scope for this task). But the backend doesn't say whether something is allowed or not, because technically it's all allowed. It just gives the frontend enough info to decide what should be done.

Reply:
This sounds like a rather great idea. I took this a step further and reused the `GET assets/<course_key>`